pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyConnection.java

```diff
@@ -481,6 +504,36 @@ public HAProxyMessage getHAProxyMessage() {
         return haProxyMessage;
     }
 
-    private static final Logger LOG = LoggerFactory.getLogger(ProxyConnection.class);
+    private boolean isBrokerActive(String targetBrokerHostPort) {
+        for (ServiceLookupData serviceLookupData : getAvailableBrokers()) {
+            if (matchesHostAndPort("pulsar://", serviceLookupData.getPulsarServiceUrl(), targetBrokerHostPort)
+                    || matchesHostAndPort("pulsar+ssl://", serviceLookupData.getPulsarServiceUrlTls(),
+                    targetBrokerHostPort)) {
+                return true;
+            }
+        }
+        return false;
+    }
 
+    private List<? extends ServiceLookupData> getAvailableBrokers() {
+        if (service.getDiscoveryProvider() == null) {
+            LOG.warn("Unable to retrieve active brokers. service.getDiscoveryProvider() is null."
+                    + "zookeeperServers and configurationStoreServers must be configured in proxy configuration "
+                    + "when checkActiveBrokers is enabled.");
+            return Collections.emptyList();
+        }
```

Comment:
@merlimat Here's the logging I added to address the issue when ZK isn't configured.

Reply:
I think a better option would be to periodically check list of brokers from REST API.

Reply:
@merlimat The REST API for getting active brokers doesn't provide sufficient information.

Here's an example call in my local test cluster:
```
❯ curl -s pulsar-testenv-deployment-broker.pulsar-testenv.svc.cluster.local:8080/admin/v2/brokers/pulsar | jq .
[
  "pulsar-testenv-deployment-broker-0.pulsar-testenv-deployment-broker.pulsar-testenv.svc.cluster.local:8080"
]
```
Implementation is [here](https://github.com/apache/pulsar/blob/af43e87d9a3bd9f18e56439521bab253cfc9ca40/pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/BrokersBase.java#L75-L103).

Is there a REST API that would return similar information as [MetadataStoreCacheLoader.getAvailableBrokers](https://github.com/apache/pulsar/blob/adcbe0f118ece0999b8603f37010194b44c241b4/pulsar-broker-common/src/main/java/org/apache/pulsar/broker/resources/MetadataStoreCacheLoader.java#L86-L95)? The [ServiceLookupData information for pulsarServiceUrl and pulsarServiceUrlTls](https://github.com/apache/pulsar/blob/301a76435ce918520f1b70b95b7ebed9bf1b08b5/pulsar-client-admin-api/src/main/java/org/apache/pulsar/policies/data/loadbalancer/ServiceLookupData.java#L32-L34 
) would be needed.

Reply:
Irrespective of the actual mechanism, we cannot just print a warning when the system is absolutely not going to work (eg: returning empty list here will deny all connections).

Reply:
> Irrespective of the actual mechanism, we cannot just print a warning when the system is absolutely not going to work

I agree. I guess that the Proxy should fail-fast at startup time if it's not properly configured?

Reply:
+1 for failing fast on start up--that will make it clear to users that it is not configured properly.

In k8s, it is easy to configure a service as headless, so that it resolves to a set of IP addresses https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#a-aaaa-records. I believe this is possible for all DNS providers, too. Would it be possible to add a feature to the proxy to consume the set of IPs to make the list of "available brokers" equivalent to the result of resolving the broker DNS entry?

I see that https://pulsar.apache.org/docs/en/administration-proxy/#use-broker-urls already recommends using a DNS entry that points to all brokers. I haven't looked closely, but given that the helm chart is not configured to use a headless service, I assume that the proxy is not able to consume a set of IPs yet.

Also, by using a headless service, we'd eliminate an additional network proxy hop.

Reply:
@lhotari Fail-fast is good, though the current issue is that `checkActiveBrokers=true` by default will break all upgrades in proxies where `zookeeperServers` is not set (which I expect to be the majority).

Reply:
> @lhotari Fail-fast is good, though the current issue is that `checkActiveBrokers=true` by default will break all upgrades in proxies where `zookeeperServers` is not set (which I expect to be the majority).

yes, I agree that breaking proxies where zookeeperServers` is not set isn't a good idea. I'll address that in follow-up changes to this PR.

Reply:
> @lhotari Fail-fast is good, though the current issue is that `checkActiveBrokers=true` by default will break all upgrades in proxies where `zookeeperServers` is not set (which I expect to be the majority).

@merlimat I have addressed this now. PTAL